File: go/pkg/hey/messages.go
package hey

import (
	"context"
	"fmt"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/basecamp/hey-sdk/go/pkg/generated"
)

// MessagesService handles message operations.
type MessagesService struct {
	client *Client
}

// NewMessagesService creates a new MessagesService.
func NewMessagesService(client *Client) *MessagesService {
	return &MessagesService{client: client}
}

// Get returns a specific message by ID.
func (s *MessagesService) Get(ctx context.Context, messageID int64) (result *generated.Message, err error) {
	op := OperationInfo{
		Service: "Messages", Operation: "GetMessage",
		ResourceType: "message", IsMutation: false, ResourceID: messageID,
	}
	if gater, ok := s.client.hooks.(GatingHooks); ok {
		if ctx, err = gater.OnOperationGate(ctx, op); err != nil {
			return
		}
	}
	start := time.Now()
	ctx = s.client.hooks.OnOperationStart(ctx, op)
	defer func() { s.client.hooks.OnOperationEnd(ctx, op, err, time.Since(start)) }()

	s.client.initGeneratedClient()
	resp, err := s.client.gen.GetMessageWithResponse(ctx, messageID)
	if err != nil {
		return nil, err
	}
	if err = CheckResponse(resp.HTTPResponse); err != nil {
		return nil, err
	}
	return resp.JSON200, nil
}

// Create creates a new message (starts a new thread) and delivers it.
// The acting sender ID is automatically resolved.
//
// Wire format (MessagesController#create): {acting_sender_id, message: {subject, content},
// entry: {addressed: {directly: [...], copied: [...], blindcopied: [...]}}}.
// Recipient lists are JSON arrays; haystack applies Array() to each kind.
func (s *MessagesService) Create(ctx context.Context, subject, content string, to, cc, bcc []string) (err error) {
	if len(to)+len(cc)+len(bcc) == 0 {
		return ErrUsage("a message needs at least one recipient (to, cc or bcc)")
	}
	op := OperationInfo{
		Service: "Messages", Operation: "CreateMessage",
		ResourceType: "message", IsMutation: true,
	}
	if gater, ok := s.client.hooks.(GatingHooks); ok {
		if ctx, err = gater.OnOperationGate(ctx, op); err != nil {
			return
		}
	}
	start := time.Now()
	ctx = s.client.hooks.OnOperationStart(ctx, op)
	defer func() { s.client.hooks.OnOperationEnd(ctx, op, err, time.Since(start)) }()

	senderID, err := s.client.DefaultSenderID(ctx)
	if err != nil {
		return err
	}

	body := generated.CreateMessageRequestContent{
		ActingSenderId: senderID,
		Message:        generated.MessagePayload{Subject: subject, Content: content},
		Entry:          entryPayload(to, cc, bcc),
	}

	resp, err := s.client.genClient().CreateMessageWithResponse(ctx, body)
	if err != nil {
		return err
	}
	return CheckResponse(resp.HTTPResponse)
}

// entryPayload builds entry.addressed for a message or reply. Callers guarantee at
// least one recipient; it still returns nil for none so the "entry" key is omitted
// rather than sent as an empty hash.
func entryPayload(to, cc, bcc []string) *generated.MessageEntryPayload {
	if len(to)+len(cc)+len(bcc) == 0 {
		return nil
	}
	return &generated.MessageEntryPayload{
		Addressed: &generated.MessageAddressed{Directly: to, Copied: cc, Blindcopied: bcc},
	}
}

// DraftContent is the whole of what a draft carries: the subject, the Trix HTML body,
// the recipients per kind and any scheduled delivery. HEY revises a draft from the
// whole of it — see UpdateDraft — so a caller edits by reading the draft (GetEdit),
// changing fields and sending everything back.
type DraftContent struct {
	Subject string
	Content string
	To      []string
	CC      []string
	BCC     []string

	// Schedule delivers the draft at an hour of a day. Nil means no scheduled
	// delivery — and on an update, clears one already set.
	Schedule *DraftSchedule
}

// DraftSchedule names a delivery time to the hour, read in the identity's time zone.
// HEY schedules to the hour; there are no minutes.
type DraftSchedule struct {
	// Date is YYYY-MM-DD, "today" or "tomorrow".
	Date string
	// Hour is 0 through 23.
	Hour int
}

// entryStatusDrafted is what entry.status carries to keep an entry a draft. Any other
// value — or omitting the status — has the server deliver the entry.
const entryStatusDrafted = "drafted"

// draftEntryPayload builds the entry payload that keeps a message drafted. Unlike
// entryPayload it always carries addressed, empty lists included: a draft with no
// recipients yet is the normal case, and on an update an empty addressed is how
// recipients are removed (HEY replaces the recipient set with what is sent).
func draftEntryPayload(draft DraftContent) *generated.MessageEntryPayload {
	entry := &generated.MessageEntryPayload{
		Status:    entryStatusDrafted,
		Addressed: &generated.MessageAddressed{Directly: draft.To, Copied: draft.CC, Blindcopied: draft.BCC},
	}
	if draft.Schedule != nil {
		entry.ScheduledDelivery = "true"
		entry.ScheduledDeliveryAtDate = draft.Schedule.Date
		entry.ScheduledDeliveryAtHour = strconv.Itoa(draft.Schedule.Hour)
	}
	return entry
}

// draftEntryIDFromLocation reads the draft's entry id out of the Location header a
// draft save answers with (204 No Content, Location: …/messages/{entry_id}) — the
// draft path serves no body, so the header is the only place the id is named.
func draftEntryIDFromLocation(resp *http.Response) (int64, error) {
	location := resp.Header.Get("Location")
	if location == "" {
		return 0, fmt.Errorf("draft saved but the response named no Location; cannot report the draft's id")
	}
	parsed, err := url.Parse(location)
	if err != nil {
		return 0, fmt.Errorf("draft saved but its Location %q is unreadable: %w", location, err)
	}
	segments := strings.Split(strings.Trim(parsed.Path, "/"), "/")
	id, err := strconv.ParseInt(segments[len(segments)-1], 10, 64)
	if err != nil || id <= 0 {
		return 0, fmt.Errorf("draft saved but its Location %q names no entry id", location)
	}
	return id, nil
}

// CreateDraft saves a new message as a draft instead of delivering it, and answers the
// draft's entry id — the id GetEdit, UpdateDraft, SendDraft and DeleteDraft take. A
// draft needs no recipients; whatever it carries is kept for the send.
func (s *MessagesService) CreateDraft(ctx context.Context, draft DraftContent) (entryID int64, err error) {
	op := OperationInfo{
		Service: "Messages", Operation: "CreateMessage",
		ResourceType: "draft", IsMutation: true,
	}

	err = s.client.instrument(ctx, op, func(ctx context.Context) error {
		senderID, serr := s.client.DefaultSenderID(ctx)
		if serr != nil {
			return serr
		}

		body := generated.CreateMessageRequestContent{
			ActingSenderId: senderID,
			Message:        generated.MessagePayload{Subject: draft.Subject, Content: draft.Content},
			Entry:          draftEntryPayload(draft),
		}

		resp, rerr := s.client.genClient().CreateMessageWithResponse(ctx, body)
		if rerr != nil {
			return rerr
		}
		if cerr := CheckResponse(resp.HTTPResponse); cerr != nil {
			return cerr
		}
		entryID, rerr = draftEntryIDFromLocation(resp.HTTPResponse)
		return rerr
	})
	return entryID, err
}

// UpdateDraft revises a draft in place from the whole of draft: the subject, the body
// and the recipients are replaced with what is sent (empty recipients remove them),
// and the scheduled delivery is rewritten too — a nil Schedule clears one already set.
// A trashed draft is silently restored by the revision.
func (s *MessagesService) UpdateDraft(ctx context.Context, entryID int64, draft DraftContent) error {
	op := OperationInfo{
		Service: "Messages", Operation: "UpdateMessage",
		ResourceType: "draft", IsMutation: true, ResourceID: entryID,
	}

	return s.client.instrument(ctx, op, func(ctx context.Context) error {
		senderID, serr := s.client.DefaultSenderID(ctx)
		if serr != nil {
			return serr
		}

		body := generated.CreateMessageRequestContent{
			ActingSenderId: senderID,
			Message:        generated.MessagePayload{Subject: draft.Subject, Content: draft.Content},
			Entry:          draftEntryPayload(draft),
		}

		resp, rerr := s.client.genClient().UpdateMessageWithResponse(ctx, entryID, body)
		if rerr != nil {
			return rerr
		}
		return CheckResponse(resp.HTTPResponse)
	})
}

// SendDraft delivers a draft through HEY's undo-delay window. The revision and the
// delivery are one request, so the draft's final state rides along: subject, body and
// recipients are replaced with what is sent, exactly as UpdateDraft replaces them.
// Delivery needs somebody to deliver to, so at least one recipient is required.
//
// The request is never retried, despite the PUT: it triggers a delivery, and a retry
// after an ambiguous first attempt could send the message twice. An ambiguous failure
// is the caller's to resolve — read the draft (or the thread) before trying again.
func (s *MessagesService) SendDraft(ctx context.Context, entryID int64, draft DraftContent) error {
	if len(draft.To)+len(draft.CC)+len(draft.BCC) == 0 {
		return ErrUsage("sending a draft needs at least one recipient (to, cc or bcc)")
	}
	op := OperationInfo{
		Service: "Messages", Operation: "UpdateMessage",
		ResourceType: "message", IsMutation: true, ResourceID: entryID,
	}

	return s.client.instrument(ctx, op, func(ctx context.Context) error {
		senderID, serr := s.client.DefaultSenderID(ctx)
		if serr != nil {
			return serr
		}

		body := generated.CreateMessageRequestContent{
			ActingSenderId: senderID,
			Message:        generated.MessagePayload{Subject: draft.Subject, Content: draft.Content},
			Entry:          entryPayload(draft.To, draft.CC, draft.BCC),
		}

		resp, rerr := s.client.genClient().UpdateMessageWithResponse(ctx, entryID, body)
		if rerr != nil {
			return rerr
		}
		return CheckResponse(resp.HTTPResponse)
	})
}

// GetEdit answers a draft's editable state — the subject, the Trix HTML body, the
// recipients per kind and any scheduled delivery, as the composer would load them.
func (s *MessagesService) GetEdit(ctx context.Context, entryID int64) (result *generated.MessageEditState, err error) {
	op := OperationInfo{
		Service: "Messages", Operation: "GetMessageEdit",
		ResourceType: "draft", IsMutation: false, ResourceID: entryID,
	}

	err = s.client.instrument(ctx, op, func(ctx context.Context) error {
		resp, rerr := s.client.genClient().GetMessageEditWithResponse(ctx, entryID)
		if rerr != nil {
			return rerr
		}
		if cerr := CheckResponse(resp.HTTPResponse); cerr != nil {
			return cerr
		}
		result = resp.JSON200
		return nil
	})
	return result, err
}

File: rust/hey-sdk/src/services/messages.rs
//! Delivering messages and keeping drafts, on top of the generated message routes.

use crate::client::Response;
use crate::error::Error;
use crate::generated::routes;
use crate::generated::types::{
    CreateMessageRequestContent, MessageAddressed, MessageEntryPayload, MessagePayload,
};

pub use crate::generated::services::messages::*;

/// What `entry.status` carries to keep an entry a draft. Any other value — or omitting
/// the status — has HEY deliver it.
const DRAFTED: &str = "drafted";

/// A message to deliver: the subject, the Trix HTML body and the recipients per kind.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct MessageContent {
    pub subject: String,
    pub content: String,
    pub to: Vec<String>,
    pub cc: Vec<String>,
    pub bcc: Vec<String>,
    /// The identity the message goes out as. `None` resolves the client's default sender.
    pub acting_sender_id: Option<i64>,
}

/// When a draft goes out, read in the identity's time zone. HEY schedules to the hour.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct DeliverySchedule {
    /// `YYYY-MM-DD`, "today" or "tomorrow".
    pub date: String,
    /// 0 through 23.
    pub hour: u8,
}

/// The whole of what a draft carries. HEY revises a draft from the whole of it, so a
/// caller edits by reading the draft, changing fields and sending everything back: empty
/// recipients remove them, and a `None` schedule clears one already set.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct DraftContent {
    pub subject: String,
    pub content: String,
    pub to: Vec<String>,
    pub cc: Vec<String>,
    pub bcc: Vec<String>,
    /// The identity the draft is saved — and ultimately delivered — as. `None` resolves
    /// the client's default sender. A caller who chose an alternate identity carries it
    /// on every revision, since a revision that leaves it out hands the draft back to
    /// the default one.
    pub acting_sender_id: Option<i64>,
    pub schedule: Option<DeliverySchedule>,
}

impl<'a> Messages<'a> {
    /// Delivers a new message through HEY's undo-delay window. Delivery needs somebody to
    /// deliver to, so at least one recipient is required.
    pub async fn send(&self, message: &MessageContent) -> Result<(), Error> {
        if !has_recipients(&message.to, &message.cc, &message.bcc) {
            return Err(Error::usage(
                "a message needs at least one recipient (to, cc or bcc)",
            ));
        }

        let body = CreateMessageRequestContent {
            acting_sender_id: self.sender_for(message.acting_sender_id).await?,
            message: MessagePayload {
                subject: message.subject.clone(),
                content: message.content.clone(),
            },
            entry: Some(delivered_entry(&message.to, &message.cc, &message.bcc)),
        };
        let mut operation = self.client().operation(&routes::CREATE_MESSAGE, &[]);
        operation.json(&body)?;
        self.client().send_unit(operation).await
    }

    /// Saves a new message as a draft instead of delivering it, and answers the draft's
    /// entry id — the id [`Messages::get_edit`], [`Messages::update_draft`],
    /// [`Messages::send_draft`] and `Entries::delete_draft` take. A draft needs no
    /// recipients; whatever it carries is kept for the send.
    pub async fn create_draft(&self, draft: &DraftContent) -> Result<i64, Error> {
        let body = self.drafted_request(draft).await?;
        let mut operation = self.client().operation(&routes::CREATE_MESSAGE, &[]);
        operation.json(&body)?;
        let response = self.client().execute(operation).await?;
        entry_id_from_location(&response)
    }

    /// Revises a draft in place from the whole of `draft`. A trashed draft is silently
    /// restored by the revision.
    pub async fn update_draft(&self, entry_id: i64, draft: &DraftContent) -> Result<(), Error> {
        let body = self.drafted_request(draft).await?;
        let mut operation = self
            .client()
            .operation(&routes::UPDATE_MESSAGE, &[&entry_id]);
        operation.json(&body)?;
        self.client().send_unit(operation).await
    }

    /// Delivers a draft through HEY's undo-delay window. The revision and the delivery are
    /// one request, so the draft's final state rides along: subject, body and recipients
    /// are replaced with what is sent, exactly as [`Messages::update_draft`] replaces them.
    ///
    /// The request is never retried, despite the PUT: it triggers a delivery, and a resend
    /// after an ambiguous first attempt could send the message twice. Reading the draft, or
    /// the thread, is the caller's way out of an ambiguous failure.
    pub async fn send_draft(&self, entry_id: i64, draft: &DraftContent) -> Result<(), Error> {
        if !has_recipients(&draft.to, &draft.cc, &draft.bcc) {
            return Err(Error::usage(
                "sending a draft needs at least one recipient (to, cc or bcc)",
            ));
        }

        let body = CreateMessageRequestContent {
            acting_sender_id: self.sender_for(draft.acting_sender_id).await?,
            message: MessagePayload {
                subject: draft.subject.clone(),
                content: draft.content.clone(),
            },
            entry: Some(delivered_entry(&draft.to, &draft.cc, &draft.bcc)),
        };
        let mut operation = self
            .client()
            .operation(&routes::UPDATE_MESSAGE, &[&entry_id]);
        operation.json(&body)?;
        operation.idempotent(false);
        self.client().send_unit(operation).await
    }

    async fn sender_for(&self, chosen: Option<i64>) -> Result<i64, Error> {
        match chosen {
            Some(id) => Ok(id),
            None => self.client().default_sender_id().await,
        }
    }

    async fn drafted_request(
        &self,
        draft: &DraftContent,
    ) -> Result<CreateMessageRequestContent, Error> {
        let mut entry = drafted_entry(&draft.to, &draft.cc, &draft.bcc);
        if let Some(schedule) = &draft.schedule {
            entry.scheduled_delivery = Some("true".to_string());
            entry.scheduled_delivery_at_date = Some(schedule.date.clone());
            entry.scheduled_delivery_at_hour = Some(schedule.hour.to_string());
        }
        Ok(CreateMessageRequestContent {
            acting_sender_id: self.sender_for(draft.acting_sender_id).await?,
            message: MessagePayload {
                subject: draft.subject.clone(),
                content: draft.content.clone(),
            },
            entry: Some(entry),
        })
    }
}

pub(crate) fn has_recipients(to: &[String], cc: &[String], bcc: &[String]) -> bool {
    !to.is_empty() || !cc.is_empty() || !bcc.is_empty()
}

/// The entry of a message HEY delivers, whose recipient kinds are the ones that name
/// somebody.
pub(crate) fn delivered_entry(to: &[String], cc: &[String], bcc: &[String]) -> MessageEntryPayload {
    let addressed = MessageAddressed {
        directly: recipients(to),
        copied: recipients(cc),
        blindcopied: recipients(bcc),
    };
    MessageEntryPayload {
        addressed: Some(addressed),
        ..MessageEntryPayload::default()
    }
}

fn recipients(addresses: &[String]) -> Option<Vec<String>> {
    if addresses.is_empty() {
        None
    } else {
        Some(addresses.to_vec())
    }
}

/// The entry of a message HEY keeps as a draft. Every recipient kind is present, empty
/// ones included: a draft addressed to nobody yet is the normal case, and on a revision
/// an empty list is how recipients are removed.
pub(crate) fn drafted_entry(to: &[String], cc: &[String], bcc: &[String]) -> MessageEntryPayload {
    let addressed = MessageAddressed {
        directly: Some(to.to_vec()),
        copied: Some(cc.to_vec()),
        blindcopied: Some(bcc.to_vec()),
    };
    MessageEntryPayload {
        addressed: Some(addressed),
        status: Some(DRAFTED.to_string()),
        ..MessageEntryPayload::default()
    }
}

/// The entry id out of the `Location` a draft save answers with: the save serves no body,
/// so the header is the only place the id is named.
pub(crate) fn entry_id_from_location(response: &Response) -> Result<i64, Error> {
    let status = response.status.as_u16();
    let location = response.header("location").ok_or_else(|| {
        Error::api(
            status,
            "draft saved but the response named no Location; cannot report the draft's id",
        )
    })?;
    let url = response.url.join(location).map_err(|error| {
        Error::api(
            status,
            format!("draft saved but its Location {location:?} is unreadable: {error}"),
        )
    })?;
    match url.path().rsplit('/').next().unwrap_or_default().parse() {
        Ok(entry_id) if entry_id > 0 => Ok(entry_id),
        _ => Err(Error::api(
            status,
            format!("draft saved but its Location {location:?} names no entry id"),
        )),
    }
}

#[cfg(test)]
mod tests {
    use std::time::Duration;

    use serde_json::{Value, json};
    use wiremock::matchers::{method, path};
    use wiremock::{Mock, MockServer, ResponseTemplate};

    use super::*;
    use crate::auth::StaticTokenProvider;
    use crate::client::Client;
    use crate::config::Config;
    use crate::error::ErrorCode;

    #[tokio::test]
    async fn send_refuses_a_message_addressed_to_nobody() {
        let server = MockServer::start().await;
        let message = MessageContent {
            subject: "Hello".to_string(),
            content: "Body".to_string(),
            ..MessageContent::default()
        };

        let error = client(&server).messages().send(&message).await.unwrap_err();

        assert_eq!(error.code(), ErrorCode::Usage);
        assert!(server.received_requests().await.unwrap().is_empty());
    }

    #[tokio::test]
    async fn send_resolves_the_identity_default_sender() {
        let server = MockServer::start().await;
        let identity = json!({ "id": 1, "senders": [{ "id": 41 }, { "id": 42, "default": true }] });
        Mock::given(method("GET"))
            .and(path("/identity.json"))
            .respond_with(ResponseTemplate::new(200).set_body_json(identity))
            .mount(&server)
            .await;
        Mock::given(method("POST"))
            .and(path("/messages.json"))
            .respond_with(ResponseTemplate::new(204))
            .mount(&server)
            .await;
        let message = MessageContent {
            subject: "Hello".to_string(),
            content: "Body".to_string(),
            to: vec!["someone@example.com".to_string()],
            ..MessageContent::default()
        };

        client(&server).messages().send(&message).await.unwrap();

        let requests = server.received_requests().await.unwrap();
        assert_eq!(requests.len(), 2);
        assert_eq!(requests[0].url.path(), "/identity.json");
        let body: Value = serde_json::from_slice(&requests[1].body).unwrap();
        assert_eq!(body["acting_sender_id"], 42);
        assert_eq!(
            body["message"],
            json!({ "subject": "Hello", "content": "Body" })
        );
        assert_eq!(
            body["entry"],
            json!({ "addressed": { "directly": ["someone@example.com"] } })
        );
    }

    #[tokio::test]
    async fn create_draft_saves_it_drafted_and_answers_the_entry_id() {
        let server = MockServer::start().await;
        Mock::given(method("POST"))
            .and(path("/messages.json"))
            .respond_with(
                ResponseTemplate::new(204)
                    .insert_header("Location", "https://app.hey.com/messages/777"),
            )
            .mount(&server)
            .await;
        let draft = DraftContent {
            subject: "From the support address".to_string(),
            content: "Draft body".to_string(),
            acting_sender_id: Some(314),
            schedule: Some(DeliverySchedule {
                date: "2026-09-02".to_string(),
                hour: 0,
            }),
            ..DraftContent::default()
        };

        let entry_id = client(&server)
            .messages()
            .create_draft(&draft)
            .await
            .unwrap();

        assert_eq!(entry_id, 777);
        let body = sent_json(&server).await;
        assert_eq!(body["acting_sender_id"], 314);
        assert_eq!(
            body["entry"],
            json!({
                "addressed": { "directly": [], "copied": [], "blindcopied": [] },
                "status": "drafted",
                "scheduled_delivery": "true",
                "scheduled_delivery_at_date": "2026-09-02",
                "scheduled_delivery_at_hour": "0"
            })
        );
    }

    #[tokio::test]
    async fn update_draft_says_the_acting_sender_back() {
        let server = MockServer::start().await;
        Mock::given(method("PUT"))
            .and(path("/messages/777.json"))
            .respond_with(ResponseTemplate::new(204))
            .mount(&server)
            .await;
        let draft = DraftContent {
            subject: "From the support address (v2)".to_string(),
            content: "Rewritten body".to_string(),
            to: vec!["someone@example.com".to_string()],
            acting_sender_id: Some(314),
            ..DraftContent::default()
        };

        client(&server)
            .messages()
            .update_draft(777, &draft)
            .await
            .unwrap();

        let body = sent_json(&server).await;
        assert_eq!(body["acting_sender_id"], 314);
        assert_eq!(body["entry"]["status"], "drafted");
        assert_eq!(
            body["entry"]["addressed"],
            json!({ "directly": ["someone@example.com"], "copied": [], "blindcopied": [] })
        );
        assert!(body["entry"]["scheduled_delivery"].is_null());
    }

    #[tokio::test]
    async fn send_draft_delivers_it_by_leaving_the_status_off() {
        let server = MockServer::start().await;
        Mock::given(method("PUT"))
            .and(path("/messages/777.json"))
            .respond_with(ResponseTemplate::new(200).set_body_json(json!({ "id": 777 })))
            .mount(&server)
            .await;

        client(&server)
            .messages()
            .send_draft(777, &deliverable_draft())
            .await
            .unwrap();

        let body = sent_json(&server).await;
        assert_eq!(body["acting_sender_id"], 314);
        assert!(body["entry"]["status"].is_null());
        assert_eq!(
            body["entry"]["addressed"],
            json!({ "directly": ["someone@example.com"] })
        );
    }

    #[tokio::test]
    async fn send_draft_refuses_a_draft_addressed_to_nobody() {
        let server = MockServer::start().await;
        let draft = DraftContent {
            to: Vec::new(),
            ..deliverable_draft()
        };

        let error = client(&server)
            .messages()
            .send_draft(777, &draft)
            .await
            .unwrap_err();

        assert_eq!(error.code(), ErrorCode::Usage);
        assert!(server.received_requests().await.unwrap().is_empty());
    }

    #[tokio::test]
    async fn send_draft_is_never_resent() {
        let server = MockServer::start().await;
        Mock::given(method("PUT"))
            .and(path("/messages/777.json"))
            .respond_with(ResponseTemplate::new(503))
            .mount(&server)
            .await;
        let client = Client::builder(Config::default().with_base_url(server.uri()))
            .token_provider(StaticTokenProvider::new("t"))
            .base_delay(Duration::from_millis(1))
            .build()
            .unwrap();

        let error = client
            .messages()
            .send_draft(777, &deliverable_draft())
            .await
            .unwrap_err();

        assert_eq!(error.http_status(), Some(503));
        assert_eq!(server.received_requests().await.unwrap().len(), 1);
    }

    #[tokio::test]
    async fn a_draft_save_without_a_location_is_an_error() {
        let server = MockServer::start().await;
        Mock::given(method("POST"))
            .and(path("/messages.json"))
            .respond_with(ResponseTemplate::new(204))
            .mount(&server)
            .await;

        let draft = DraftContent {
            acting_sender_id: Some(314),
            ..DraftContent::default()
        };

        let error = client(&server)
            .messages()
            .create_draft(&draft)
            .await
            .unwrap_err();

        assert_eq!(error.code(), ErrorCode::Api);
    }

    fn deliverable_draft() -> DraftContent {
        DraftContent {
            subject: "From the support address".to_string(),
            content: "Final body".to_string(),
            to: vec!["someone@example.com".to_string()],
            acting_sender_id: Some(314),
            ..DraftContent::default()
        }
    }

    fn client(server: &MockServer) -> Client {
        Client::builder(Config::default().with_base_url(server.uri()))
            .token_provider(StaticTokenProvider::new("t"))
            .max_retries(0)
            .build()
            .unwrap()
    }

    async fn sent_json(server: &MockServer) -> Value {
        let requests = server.received_requests().await.unwrap();
        assert_eq!(requests.len(), 1);
        serde_json::from_slice(&requests[0].body).unwrap()
    }
}
